Continuous testing fails in multi-module project when using module as a dependency in 3.22 onwards

### Describe the bug

When running continuous testing on a module within a multi-module project that uses one of the other modules as a dependency (in this case an internal-model module which doesn't contain a service), the continuous testing will fail. It seems a prod config profile is created for the imported module despite it not being a service, which causes issues when trying to find the TestConfig. It can be resolved by using the property `quarkus.test.only-test-application-module=true`, but this wasn't required prior to 3.22

Stacktrace:

> 2025-06-20 16:24:08,492 ERROR [io.qua.test] (Test runner thread) ==================== TEST REPORT #1 ====================
> 2025-06-20 16:24:08,492 ERROR [io.qua.test] (Test runner thread) Test DataTest failed 
> : org.junit.jupiter.engine.execution.ConditionEvaluationException: Failed to evaluate condition [io.quarkus.test.junit.QuarkusTestExtension]: Internal error: Test class was loaded with an unexpected classloader or the thread context classloader was incorrect.
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
> 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
> 	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
> 	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1685)
> 	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
> 	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
> 	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
> 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
> 	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> Caused by: java.lang.IllegalStateException: Internal error: Test class was loaded with an unexpected classloader or the thread context classloader was incorrect.
> 	at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:201)
> 	... 14 more
> Caused by: java.util.NoSuchElementException: SRCFG00027: Could not find a mapping for io.quarkus.deployment.dev.testing.TestConfig
> 	at io.smallrye.config.SmallRyeConfig.getConfigMapping(SmallRyeConfig.java:647)
> 	at io.smallrye.config.SmallRyeConfig.getConfigMapping(SmallRyeConfig.java:637)
> 	at io.quarkus.test.junit.AbstractJvmQuarkusTestExtension.evaluateExecutionCondition(AbstractJvmQuarkusTestExtension.java:161)
> 	... 14 more

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

I have reproduced the issue here: https://github.com/hcward/quarkus-cont-testing

1. The controller module doesn't import the internal-model and runs successfully
2. The data module imports the internal-mode and fails

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/dev/testing/JunitTestRunner.java
@@ -153,9 +153,7 @@ public class JunitTestRunner {
             ClassLoader old = Thread.currentThread().getContextClassLoader();
             QuarkusClassLoader tcl = testApplication.createDeploymentClassLoader();
             deploymentClassLoader = tcl;
-            if (firstDeploymentClassLoader == null) {
-                firstDeploymentClassLoader = deploymentClassLoader;
-            }
+
             LogCapturingOutputFilter logHandler = new LogCapturingOutputFilter(testApplication, true, true,
                     TestSupport.instance()
                             .get()::isDisplayTestOutput);
@@ -724,9 +722,13 @@ public class JunitTestRunner {
         ClassLoader classLoaderForLoadingTests;
         Closeable classLoaderToClose = null;
         ClassLoader orig = Thread.currentThread().getContextClassLoader();
+        // JUnitTestRunner is loaded with an augmentation classloader which does not have visibility of FacadeClassLoader, but the deployment classloader can see it
+        // We need a consistent classloader or we leak curated applications, so use a static classloader we stashed away
+        // In a multi-module project we will have several deployment classloaders, but it's ok
+        if (firstDeploymentClassLoader == null) {
+            firstDeploymentClassLoader = deploymentClassLoader;
+        }
         try {
-            // JUnitTestRunner is loaded with an augmentation classloader which does not have visibility of FacadeClassLoader, but the deployment classloader can see it
-            // We need a consistent classloader or we leak curated applications, so use a static classloader we stashed away
             Class fclClazz = firstDeploymentClassLoader.loadClass(FACADE_CLASS_LOADER_NAME);
             Constructor constructor = fclClazz.getConstructor(ClassLoader.class, boolean.class, CuratedApplication.class,
                     Map.class,
@@ -746,6 +748,8 @@ public class JunitTestRunner {
                     .setContextClassLoader(classLoaderForLoadingTests);
         } catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException | InstantiationException
                 | InvocationTargetException e) {
+            // If the first deployment classloader cannot load a facade classloader, don't keep using it, let the next module provide one
+            firstDeploymentClassLoader = null;
             // This is fine, and usually just means that test-framework/junit5 isn't one of the project dependencies
             // In that case, fallback to loading classes as we normally would, using a TCCL
             log.debug(

--- a/integration-tests/maven/src/test/resources-filtered/projects/multimodule/dependency/pom.xml
+++ b/integration-tests/maven/src/test/resources-filtered/projects/multimodule/dependency/pom.xml
@@ -1,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project>
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.acme</groupId>
+        <artifactId>quarkus-quickstart-multimodule-parent</artifactId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <groupId>org.acme</groupId>
+    <artifactId>quarkus-quickstart-multimodule-dependency</artifactId>
+    <version>1.0-SNAPSHOT</version>
+
+</project>

--- a/integration-tests/maven/src/test/resources-filtered/projects/multimodule/pom.xml
+++ b/integration-tests/maven/src/test/resources-filtered/projects/multimodule/pom.xml
@@ -20,6 +20,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     </properties>
     <modules>
+        <module>dependency</module>
         <module>rest</module>
         <module>html</module>
         <module>runner</module>
@@ -61,6 +62,11 @@
             <dependency>
                 <groupId>org.acme</groupId>
                 <artifactId>quarkus-quickstart-multimodule-rest</artifactId>
+                <version>1.0-SNAPSHOT</version>
+            </dependency>
+            <dependency>
+                <groupId>org.acme</groupId>
+                <artifactId>quarkus-quickstart-multimodule-dependency</artifactId>
                 <version>1.0-SNAPSHOT</version>
             </dependency>
         </dependencies>

--- a/integration-tests/maven/src/test/resources-filtered/projects/multimodule/rest/src/test/java/org/acme/test/SimpleTest.java
+++ b/integration-tests/maven/src/test/resources-filtered/projects/multimodule/rest/src/test/java/org/acme/test/SimpleTest.java
@@ -4,7 +4,7 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 public class SimpleTest {
-    
+
     @Test
     public void testNothing() {
         Assertions.assertTrue(true);

--- a/integration-tests/maven/src/test/resources-filtered/projects/multimodule/runner/pom.xml
+++ b/integration-tests/maven/src/test/resources-filtered/projects/multimodule/runner/pom.xml
@@ -25,6 +25,10 @@
             <artifactId>quarkus-quickstart-multimodule-rest</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.acme</groupId>
+            <artifactId>quarkus-quickstart-multimodule-dependency</artifactId>
+        </dependency>
+        <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-junit5</artifactId>
             <scope>test</scope>

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/classloading/FacadeClassLoader.java
@@ -117,6 +117,7 @@ public final class FacadeClassLoader extends ClassLoader implements Closeable {
         this(parent, false, null, null, null, System.getProperty("java.class.path"));
     }
 
+    // Also called reflectively by JUnitTestRunner
     public FacadeClassLoader(ClassLoader parent, boolean isAuxiliaryApplication, CuratedApplication curatedApplication,
             final Map<String, String> profileNames,
             final Set<String> quarkusTestClasses, final String classesPath) {
@@ -550,6 +551,7 @@ public final class FacadeClassLoader extends ClassLoader implements Closeable {
         // If the try block fails, this would be null, but there's no catch, so we'd never get to this code
         QuarkusClassLoader loader = startupAction.getClassLoader();
 
+        // Make sure that our new classloader has config on it; this is a bit of a scattergun approach to setting config, but it helps cover most paths
         Class<?> configProviderResolverClass = loader.loadClass(ConfigProviderResolver.class.getName());
 
         Class<?> testConfigProviderResolverClass = loader.loadClass(QuarkusTestConfigProviderResolver.class.getName());

